hera_cal/apply_cal.py

```diff
@@ -71,84 +78,103 @@ def recalibrate_in_place(data, data_flags, new_gains, cal_flags, old_gains=None,
-    else:
-        raise NotImplementedError('This function has not been implemented yet.')
+
+    # optionally load npz flags
     if flags_npz is not None:
         npz_flags = np.load(flags_npz)
-        uvd.flag_array = np.logical_or(npz_flags['flag_array'], uvd.flag_array)
         add_to_history += ' FLAGS_NPZ_HISTORY: ' + str(npz_flags['history']) + '\n'
-    data, data_flags = io.load_vis(uvd)
-
-    # apply band edge flags
-    for bl in data_flags.keys():
-        data_flags[bl][:, 0:flag_nchan_low] = True
-        data_flags[bl][:, data_flags[bl].shape[1] - flag_nchan_high:] = True
+        npz_flag_dc = io.load_npz_flags(flags_npz)
 
     # load new calibration solution
-    if new_calibration is None:
-        raise ValueError('Must provide a calibration solution to apply.')
-    if isinstance(new_calibration, UVCal):
-        uvc = new_calibration
-    else:
-        uvc = UVCal()
-        uvc.read_calfits(new_calibration)
-    add_to_history += ' NEW_CALFITS_HISTORY: ' + uvc.history + '\n'
-    new_gains, new_flags = io.load_cal(uvc)
+    hc = io.HERACal(new_calibration)
+    new_gains, new_flags, _, _ = hc.read()
+    add_to_history += ' NEW_CALFITS_HISTORY: ' + hc.history + '\n'
 
     # load old calibration solution
     if old_calibration is not None:
-        if isinstance(old_calibration, UVCal):
-            old_uvc = old_calibration
-        else:
-            old_uvc = UVCal()
-            old_uvc.read_calfits(old_calibration)
-        add_to_history += ' OLD_CALFITS_HISTORY: ' + old_uvc.history + '\n'
-        old_calibration, _ = io.load_cal(old_uvc)
+        old_hc = io.HERACal(old_calibration)
+        old_calibration, _, _, _ = old_hc.read()
+        add_to_history += ' OLD_CALFITS_HISTORY: ' + old_hc.history + '\n'
 
-    recalibrate_in_place(data, data_flags, new_gains, new_flags, old_gains=old_calibration, gain_convention=gain_convention)
-    io.update_vis(data_infilename, data_outfilename, filetype_in=filetype, filetype_out=filetype, data=data,
-                  flags=data_flags, add_to_history=add_to_history, clobber=clobber, **kwargs)
+    # partial loading and writing using uvh5
+    if nbl_per_load is not None:
+        if not (filetype_in is 'uvh5') or not (filetype_out is 'uvh5'):
+            raise NotImplementedError('Partial writing is not implemented for non-uvh5 I/O.')
+        hd = io.HERAData(data_infilename, filetype='uvh5')
+        for attribute, value in kwargs.items():
+            hd.__setattr__(attribute, value)
+        for data, data_flags, _ in hd.iterate_over_bls(Nbls=nbl_per_load):
+            for bl in data_flags.keys():
+                # apply band edge flags
+                data_flags[bl][:, 0:flag_nchan_low] = True
+                data_flags[bl][:, data_flags[bl].shape[1] - flag_nchan_high:] = True
+                # apply npz flags
+                if flags_npz is not None:
+                    data_flags[bl] = np.logical_or(data_flags[bl], npz_flag_dc[bl])
```

Comment:
Should a test be done to see if `npz_flag_dc` has flags corresponding to `bl`? Basically, could this cause a `KeyError`?

Reply:
Yes this could cause a KeyError. I honestly think that's fine. I'll add a clarification to the flags_npz docstring.

Reply:
Okay, it'd be nice to have a more graceful failure mode (like maybe it raises a warning but keeps going), but we can keep it this way if you want.